model/cachedfeed.py

```diff
@@ -73,120 +75,233 @@ class CachedFeed(Base):
+        """Retrieve a cached feed from the database if possible.
+
+        Generate it from scratch and store it in the database if
+        necessary.
+
+        :param _db: A database connection.
+        :param worklist: The WorkList associated with this feed.
+        :param facets: A Facets object that distinguishes this feed from
+            others (for instance, by its sort order).
+        :param pagination: A Pagination object that explains which
+            page of a larger feed is being cached.
+        :param refresher_method: A function to call if it turns out
+            the contents of the feed need to be regenerated. This
+            function must take no arguments and return an object that
+            implements __unicode__. (A Unicode string or an OPDSFeed is fine.)
+        :param max_age: If a cached feed is older than this, it will
+            be considered stale and regenerated. This may be either a
+            number of seconds or a timedelta. If no value is
+            specified, a default value will be calculated based on
+            WorkList and Facets configuration. Setting this value to
+            zero will force a refresh.
+
+        :return: A CachedFeed containing up-to-date content.
+        """
 
-        if facets:
-            facets_key = unicode(facets.query_string)
-        else:
-            facets_key = u""
+        # Gather the information necessary to uniquely identify this
+        # page of this feed.
+        keys = cls._prepare_keys(_db, worklist, facets, pagination)
 
-        if pagination:
-            pagination_key = unicode(pagination.query_string)
-        else:
-            pagination_key = u""
+        # Calculate the maximum cache age, converting from timedelta
+        # to seconds if necessary.
+        max_age = cls.max_cache_age(worklist, keys.feed_type, max_age)
 
-        # Get a CachedFeed object. We will either return its .content,
-        # or update its .content.
+        # These arguments will probably be passed into get_one, and
+        # will be passed into get_one_or_create in the event of a cache
+        # miss.
+
+        # TODO: this constraint_clause might not be necessary anymore.
+        # ISTR it was an attempt to avoid race conditions, and we do a
+        # better job of that now.
         constraint_clause = and_(cls.content!=None, cls.timestamp!=None)
-        feed, is_new = get_one_or_create(
-            _db, cls,
+        kwargs = dict(
             on_multiple='interchangeable',
             constraint=constraint_clause,
-            lane_id=lane_id,
-            unique_key=unique_key,
-            library=library,
-            work=work,
-            type=type,
-            facets=facets_key,
-            pagination=pagination_key)
-
-        if force_refresh is True:
-            # No matter what, we've been directed to treat this
-            # cached feed as stale.
-            return feed, False
-
-        if max_age is lane.CACHE_FOREVER:
```

Comment:
Most of the code in this method was refactored, but this bit was removed for simplicity. We no longer need to be afraid of a grouped feed that's so expensive we just can't generate it and we have to fall back to generating a paginated feed. It's not good to generate a grouped feed while an HTTP request is happening, and in normal situations we won't do that, but it's not the end of the world -- it just means one request is slow.